p4_symbolic/symbolic/symbolic.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "p4_symbolic/symbolic/symbolic.h"
+
+#include "p4_symbolic/symbolic/table.h"
+#include "p4_symbolic/symbolic/util.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+
+z3::context *Z3_CONTEXT = nullptr;
+
+pdpi::StatusOr<SolverState *> EvaluateP4Pipeline(
+    const Dataplane &data_plane, const std::vector<int> &physical_ports) {
+  // Check Z3_CONTEXT is freed properly to avoid any context sharing issues.
+  if (Z3_CONTEXT != nullptr) {
+    return absl::Status(absl::StatusCode::kAlreadyExists,
+                        "Cannot make subsequent call to EvaluateP4Pipeline() "
+                        "when SolverState returned by a previous call has not "
+                        "been destructed!");
+  }
+
+  // Context to use for defining z3 variables, etc..
```

Comment:
Instead of initializing the context here, can we just initialize it once for the whole program. There is a lot of subtly around global variables that have destructors (go/totw/110), but the TLDR is:

z3::context& Z3Context() {
  static z3::context* context = new z3::context();
  return *context;
}

Reply:
Actually, there is no need for Z3_CONTEXT to even be a pointer.

Reply:
So if you have the Z3Context function, you just always access the context through that function. Because the context inside the function is static, it's essentially a global variable. Inside the function, the context has to be a pointer, otherwise, the object would be destructed when the function returns (I believe).

Reply:
Oh actually, the destructor will be called at the end of the program, but I believe we don't want that, due to some subtle race conditions etc as described in go/totw/110